### PR TITLE
Adding auth state ready() ax

### PR DESCRIPTION
**Pull request by AngelAngelXie**
Fixes - https://github.com/firebase/firebase-js-sdk/issues/462

Implemented authStateReady() : This method returns a promise that resolves immediately when the initial auth state is settled and currentUser is available.

Added relevant unit tests to confirm the method works as expected. 

Edited demo app to incorporate authStateReady() method when accessing current user. 
- kept activeUser(), which returns the last user OR currentUser without calling authStateReady(). This is needed for use cases where we simply want to check if there is currently a sign in user.
- added getActiveUserAfterSignIn(), which returns the last user OR the currentUser once the user has completed signing in. This incorporates authStateReady() to make sure that currentUser information is retrieved after sign in has been successful.

### CODE REVIEW

File: packages/auth/src/model/public_types.ts [hunk 1]
```diff
@@ -291,6 +291,11 @@ export interface Auth {
     error?: ErrorFn,
     completed?: CompleteFn
   ): Unsubscribe;
+  /**
+   * return a promise that resolves immediately when the initial
+   * auth state is settled and currentUser is available
+   */
+  authStateReady(): Promise<void>;
```
Comment (prameshj) [inline packages/auth/src/model/public_types.ts:298]: This will also resolve if a user signs out.. (since that will cause onAuthStateChanged to trigger as well). Let's add more more line to the comment like:

"the current user might be a valid user, or null if there is no user signed in currently."

File: packages/auth/src/core/auth/auth_impl.test.ts [hunk 2]
```diff
@@ -786,4 +786,125 @@ describe('core/auth/auth_impl', () => {
...
+  describe('AuthStateReady', () => {
+    let user: UserInternal;
+    let onAuthStateChangedCallback: sinon.SinonSpy;
+    let authStateChangedSpy: sinon.SinonSpy;
+
+    beforeEach(async () => {
+      user = testUser(auth, 'uid');
+
+      onAuthStateChangedCallback = sinon.spy();
+      auth.onAuthStateChanged(onAuthStateChangedCallback);
```
Comment (prameshj) [inline packages/auth/src/core/auth/auth_impl.test.ts:799]: The primary purpose of "onAuthStateChangedCallback" is to ensure that updateCurrentUser calls the callback registered by onAuthStateChanged?

IMO, that shouldn't be tested in the "AuthStateReady" test block. In this test block, we want to make sure that when authStateReady is called and we already updated a current user.. it should return right away without invoking "onAuthStateChanged".

TLDR, i think we should remove lines 798, 799 and related checks throughout the file.

File: packages/auth/src/core/auth/auth_impl.test.ts [hunk 3]
```diff
@@ -786,4 +786,125 @@ describe('core/auth/auth_impl', () => {
...
+      expect(authStateChangedSpy).to.be.calledOnce;
+      // expect(onAuthStateChangedCallback).to.be.calledThrice;
+    });
+
+    //case three: if user logged out more than once, promise should still be resolved with currentUser remained as null
+    it('resolves promise with currentUser remains as null when user logs out more than once', async () => {
+      expect(authStateChangedSpy).to.not.have.been.called;
+      const promiseVar = auth.authStateReady();
+      expect(authStateChangedSpy).to.be.calledOnce;
+
```
Comment (prameshj) [inline packages/auth/src/core/auth/auth_impl.test.ts:850]: nit: add comment "simulate a logout"

File: packages/auth/src/core/auth/auth_impl.test.ts [hunk 4]
```diff
@@ -786,4 +786,111 @@ describe('core/auth/auth_impl', () => {
...
+      await auth._updateCurrentUser(null);
+    });
+
+    it('immediately returns resolved promise if the user is previously logged in', async () => {
+      await auth._updateCurrentUser(user);
+
+      await auth
+        .authStateReady()
+        .then(() => {
```
Comment (prameshj) [inline packages/auth/src/core/auth/auth_impl.test.ts:807]: nit: add a comment here like:

"authStateReady calls onAuthStateChanged, if there is no currentUser available"

File: packages/auth/src/core/auth/auth_impl.test.ts [hunk 5]
```diff
@@ -786,4 +786,111 @@ describe('core/auth/auth_impl', () => {
...
+          expect(auth.currentUser).to.eq(null);
+        })
+        .catch(error => {
+          throw new Error(error);
+        });
+
+      expect(authStateChangedSpy).to.be.calledOnce;
+    });
+
+    it('resolves the promise with currentUser remain null during log in failure', async () => {
```
Comment (prameshj) [inline packages/auth/src/core/auth/auth_impl.test.ts:852]: nit: replace "remain null" with "being null"

**Review by prameshj — APPROVED** [review]

File: docs-devsite/auth.auth.md [hunk 6]
```diff
@@ -37,6 +37,7 @@ export interface Auth
 
 |  Method | Description |
 |  --- | --- |
+|  [authStateReady()](./auth.auth.md#authauthstateready) | return a promise that resolves immediately when the initial auth state is settled. The current user might be a valid user, or null if there is no user signed in currently. |
```
Comment (egilmorez) [inline docs-devsite/auth.auth.md:40]: I think this should this be "[authStateReady()](./auth.auth.md#authauthstateready) | returns" with an "s."

Comment (egilmorez) [reply docs-devsite/auth.auth.md:40]: Also, there's something off in the sentence, "The current user might be a valid user, or null if there is no user signed in currently."

Are there two possible return values, `valid` and null?  Or does it return a username?  Suggest rephrasing and using backticks if you mention literal values.

File: docs-devsite/auth.auth.md [hunk 7]
```diff
@@ -144,6 +145,19 @@ const result = await signInWithEmailAndPassword(auth, email, password);
 
 ```
 
+## Auth.authStateReady()
+
+return a promise that resolves immediately when the initial auth state is settled. The current user might be a valid user, or null if there is no user signed in currently.
```
Comment (egilmorez) [inline docs-devsite/auth.auth.md:150]: I think this should be "Returns a promise . . ."

Along with the other fixes suggested above.

File: packages/auth/demo/src/index.js [hunk 8]
```diff
@@ -114,7 +114,7 @@ const providersIcons = {
 };
 
 /**
- * Returns the active user (i.e. currentUser or lastUser).
+ * Returns active user (i.e. currentUser or lastUser).
```
Comment (egilmorez) [inline packages/auth/demo/src/index.js:117]: Suggest "Returns the active user (`currentUser` or `lastUser`)."

Always avoid "i. e."

File: packages/auth/demo/src/index.js [hunk 9]
```diff
@@ -126,63 +126,88 @@ function activeUser() {
   }
 }
 
+/**
+ * Blocks until there is a valid user
+ * then return the valid user (i.e. currentUser or lastUser).
```
Comment (egilmorez) [inline packages/auth/demo/src/index.js:131]: Use "then returns" and add fixes suggested above, thanks!

File: packages/auth/src/core/auth/auth_impl.test.ts [hunk 10]
```diff
@@ -786,4 +786,112 @@ describe('core/auth/auth_impl', () => {
...
+
+      authStateChangedSpy = sinon.spy(auth, 'onAuthStateChanged');
+
+      await auth._updateCurrentUser(null);
+    });
+
+    it('immediately returns resolved promise if the user is previously logged in', async () => {
+      await auth._updateCurrentUser(user);
+
+      // authStateReady calls onAuthStateChanged, if there is no currentUser available
```
Comment (egilmorez) [inline packages/auth/src/core/auth/auth_impl.test.ts:805]: Comments in this file are not publicly visible, iiuc.  Please correct if wrong, thanks!

File: packages/auth/src/model/public_types.ts [hunk 11]
```diff
@@ -291,6 +291,12 @@ export interface Auth {
     error?: ErrorFn,
     completed?: CompleteFn
   ): Unsubscribe;
+  /**
```
Comment (egilmorez) [inline packages/auth/src/model/public_types.ts:294]: If this is publicly visible, please apply fixes suggested for identical text above, thanks!

**Review by egilmorez — COMMENTED** [review]
Some things to look at in the comments, thanks!

**Review by egilmorez — APPROVED** [review]
Thanks!

**Review by prameshj — APPROVED** [review]
